Ch3: wrangling (#76)
UBC-DSCI/introduction-to-datascience-python: source/wrangling.md

      - `.str.split`
  - Recall and use the following operators for their

>     intended data wrangling tasks:

Reviewer: check the learning objectives for validity.
- "define what lists, series, and data frames are" -- do we do lists explicitly now, e.g.?

Author: updated the content, so now we explicitly do lists and dictionaries

      - `==`
      - `in`
      - `and`
      - `or`
      - `[]`
      - `.loc[]`
      - `.iloc[]`

> ## Data frames, series, and lists

Reviewer: Do we cover lists any more? rename this section?

> In Chapters {ref}`intro` and {ref}`reading`, *data frames* were the focus:

Reviewer: The code
```
In Chapter {ref}`blah`, we do...
```
does not typeset properly in jupyterbook. See #50 . Need to search through whole chapter and change to

```
the {ref}`blah` chapter
```

A data frame storing data regarding the population of various regions in Canada. In this example data frame, the row that corresponds to the observation for the city of Vancouver is colored yellow, and the column that corresponds to the population variable is colored blue.
```

```{code-cell} ipython3
:tags: [remove-cell]

# The following cell was removed because there is no "vector" in Python.
```

+++ {"tags": ["remove-cell"]}

> Python stores the columns of a data frame as either

Reviewer: We need to keep some amount of this text (rewritten appropriately) to say that dataframes use series as columns, otherwise students won't know why you are covering series below

Author: :+1: see the section on Data Structures in Python (starting at line 247)

> In Python, `pandas` **series** are lists. They are ordered and can be indexed.

Reviewer: "series are lists"

I kind of prefer the R wording here: "are objects that can contain one or more elements"

also series technically not a python list

maybe write the R phrasing and then say something like "(essentially, a list)"

Reviewer: also reinforce here that series are what pandas uses for each column of a df

> They are strictly 1-dimensional and can contain any data type

Reviewer: "1-dimensional" will not mean anything to a first year. less jargon

> (integers, strings, floats, etc), including a mix of them; Python

Reviewer: although it can include a mix, we should make sure to remark that it is generally good practice to store only one kind of data in a series (because all observations of one variable should be the same type)

{numref}`tab:datatype-table`.
You can create a `pandas` series using the
`pd.Series()` function.  For example, to create the series `region` as shown

> in{numref}`fig:02-series`, you can write:

Reviewer: spacing error

> A data frame storing data regarding the population of various regions in Canada. In this example data frame, the row that corresponds to the observation for the city of Vancouver is colored yellow, and the column that corresponds to the population variable is colored blue.

Reviewer 2: It would be nice if this image also highlighted what a value is, for both R and Py

![image](https://user-images.githubusercontent.com/4560057/210589818-c349c17a-2c2f-41aa-ace8-32efc130f049.png)

```{table} Basic data types in Python
:name: tab:datatype-table
| Data type             | Abbreviation | Description                                   | Example                                    |
| :-------------------- | :----------- | :-------------------------------------------- | :----------------------------------------- |
| integer               | `int`        | positive/negative whole numbers               | `42`                                       |
| floating point number | `float`      | real number in decimal form                   | `3.14159`                                  |
| boolean               | `bool`       | true or false                                 | `True`                                     |

> | string                | `str`        | text                                          | `"Can I have a cheezburger?"`              |

Reviewer: "Can I have a cheezburger" is cute but kind of out-of-place informality here

| none                  | `NoneType`   | represents no value                           | `None`                                     |
```

+++

It is important in Python to make sure you represent your data with the correct type.
Many of the `pandas` functions we use in this book treat
the various data types differently. You should use integers and float types

> (which both fall under the "numeric" umbrella type) to represent numbers and perform

Reviewer: "numeric" isn't a type in python -- this was missed from R translation

arithmetic. Strings are used to represent data that should
be thought of as "text", such as words, names, paths, URLs, and more.

> There are other basic data types in Python, such as *set*

Reviewer: We need to discuss each type that appears in the table.

Missing bool, nonetype: what are these used for?

In the text here we should also reference the type name, so `int` `float` `bool` `str` and `NoneType` after we mention them in the sentence

E.g. 

"Strings (`str`) blahblah"

| floating point number | `float`      | real number in decimal form                   | `3.14159`                                  |
| boolean               | `bool`       | true or false                                 | `True`                                     |
| string                | `str`        | text                                          | `"Hello World"`                            |
| none                  | `NoneType`   | represents no value                           | `None`                                     |
```

+++

It is important in Python to make sure you represent your data with the correct type.
Many of the `pandas` functions we use in this book treat
the various data types differently. You should use integers and float types
to represent numbers and perform arithmetic. Integers have no decimals while floats contain decimals.
Booleans are variables that can only take on one of two values: True or False.
Strings are used to represent data that should
be thought of as "text", such as words, names, paths, URLs, and more.
A `NoneType` is a special type in Python that is used to indicate no-value; this can occur when

> you have missing data.

Author: added this per comment below

> A data frame is really just series stuck together that follows two rules:

Reviewer: "is really just series stuck together"

series is its own plural here, so it's a bit awk

maybe

"A data frame is really just a list of series....."

Author: I think the concept of a "list of series" might be a bit less intuitive than "stuck together" I will try a less backward wording, but if this doesn't fit with you, I am happy to adopt your suggestion

"
A data frame is really just a collection of series that are stuck together; it follows two rules
"

1. Each element itself is a series.
2. Each element (series) must have the same length.

Not all columns in a data frame need to be of the same type.
{numref}`fig:02-dataframe` shows a data frame where

> the columns are series of different types.

Reviewer: Should take some of the text from the R version to put here at the end. Like

"But each element in one column should usually be the same type, since the values for a single variable are usually all of the same type. So even though series let you put different types in them, most common to have just one per column blahblah"

> > For example we can check the class of the Canadian languages data set,

Reviewer: the code cell below this note is not inside the note. It should be inside the note (see the R version)

Author: hmm, that doesn't easily work in juypterbook... I created #88

Reviewer 2: I think it would be nice if we used the [jupyter book admonitions](https://jupyterbook.org/en/stable/content/content-blocks.html#notes-warnings-and-other-admonitions) ("note" in this case) instead of block quotes for these section. These can also contain code as demonstrated in the link.

> `can_lang`, we worked with in the previous chapters and we see it is a `pandas.core.frame.DataFrame`.

```{code-cell} ipython3
:tags: [remove-cell]

# The functions from the `tidyverse` package that we use often give us a

> # special class of data frame called a *tibble*. Tibbles have some additional \index{tibble}

Reviewer: We should keep this whole paragraph, but reword it to be about `pandas DataFrames`

There will be no mention of a "base python dataframe" because that doesn't exist here. But it's still worth saying what the type is, discussing that dataframes provide a bunch of functions (like loc and [] and whatever).

Try to keep the R pedagogy as much as possible

Author: See new section on `### Data structures in Python` (starting on 251)

> Lists, Series and DataFrames are basic types of *data structure* in Python, which

Reviewer: The table link here is broken (it points elsewhere in the chapter...)

also, should we cover `list`, `dict`, `tuple` and `set` here? This would probably be the place to do it. Also say "while this course will focus on data frames in Pandas, python has some very useful data structures that you will often see"

Table:
Data structure |  Example |   Usage
list   | [1,"a", 4] | storing an ordered collection of elements. You can index it like `blah`.
dict | {a : 3, blah} | storing a mapping from key to value

etc etc

Author: I took a pass at this. Rather than including examples in the table, I included code examples below. See if you think this works

> For example we can check the class of the Canadian languages data set,
> `can_lang`, we worked with in the previous chapters and we see it is a `pandas.core.frame.DataFrame`.

```{code-cell} ipython3
can_lang = pd.read_csv("data/can_lang.csv")
type(can_lang)
```

### Data structures in Python

> The `Series` and `DataFrame` types are *data structure* in Python, which

Reviewer 2: ```suggestion
The `Series` and `DataFrame` types are *data structures* in Python, which
```

are core to most data analyses.
The functions from `pandas` that we use often give us back a `DataFrame`
or a `Series` depending on the operation. Because
`Series` are essentially simple `DataFrames`, we will refer
to both `DataFrames` and `Series` as "data frames" in the text.
There are other types that represent data structures in Python.
We summarize the most common ones in {numref}`tab:datastruc-table`.

```{table} Basic data structures in Python
:name: tab:datastruc-table
| Data Structure | Description |
| ---            | ----------- |
| list | An ordered collection of values that can store multiple data types at once. |
| dict | A labeled data structure where `keys` are paired with `values` |
| Series | An ordered collection of values *with labels* that can store multiple data types at once. |
| DataFrame | A labeled data structure with `Series` columns of potentially different types. |
```

A `list` is an ordered collection of values. To create a list, we put the contents of the list in between
square brackets `[]`, where each item of the list is separated by a comma. A `list` can contain values

> of different types. The example below contains 6 `str` entries.

Reviewer 2: ```suggestion
of different types. The example below contains six `str` entries.
```

> For example we can check the class of the Canadian languages data set,
> `can_lang`, we worked with in the previous chapters and we see it is a `pandas.core.frame.DataFrame`.

```{code-cell} ipython3
can_lang = pd.read_csv("data/can_lang.csv")
type(can_lang)
```

> ### Data structures in Python

Author: This may not need its own title, but I found it helpful. Feel free to remove if you think it is best to flow from the content about DataFrames above

Reviewer 2: I think this title is helpful so I would vote for keeping it.

There are many ways a tabular data set can be organized. This chapter will focus

> on introducing the **tidy data** format of organization and how to make your raw

Reviewer 2: ```suggestion
There are many ways a tabular data set can be organized. The data frames we have looked at so far
have all been using the **tidy data** format. This chapter will focus on introducing this type of
data frame format of organization and walk through how to make your raw
```

I think the start of this tidy data section is repetition of the "What is a data frame" section where we already introduced the concept of rows, columns, and variables. Maybe we can keep only one of these illustrations and start this section off by referring to the previous one as in my suggestion above?

```{index} pandas.DataFrame; melt
```

One task that is commonly performed to get data into a tidy format 
is to combine values that are stored in separate columns, 
One task that is commonly performed to get data into a tidy format
is to combine values that are stored in separate columns,
but are really part of the same variable, into one.
Data is often stored this way 
because this format is sometimes more intuitive for human readability 
Data is often stored this way
because this format is sometimes more intuitive for human readability
and understanding, and humans create data sets.

> In {numref}`fig:02-wide-to-long`, 

Reviewer: The caption on the figure that this reference is pointing to is overlapping the image itself somehow (Fig 17). No idea why, but figure shouldn't overlap its caption

Author: I have no idea why. I started #87 . It is still readable, so I don't think critical to solve at the moment

Another problem with data in this format is that we don't know what the
numbers under each year actually represent. Do those numbers represent
population size? Land area? It's not clear.
To solve both of these problems,
we can reshape this data set to a tidy data format
by creating a column called "year" and a column called
"population." This transformation&mdash;which makes the data
"longer"&mdash;is shown as the right table in
{numref}`fig:02-wide-to-long`. Note that the number of entries in our data frame
can change in this transformation. The "untidy" data has 5 rows and 3 columns for

> a total of 15 data, whereas the "tidy" data on the right has 15 rows and 2 columns

Reviewer: We don't usually refer to entries in a data frame in this class as "data" (even though that's technically correct). I would use "entries" here

```

We can achieve this effect in Python using the `.melt` function from the `pandas` package.

> We say that we "melt" (or "pivot") the wide table into a longer format.

Reviewer: I would remove this new "We say that" sentence -- especially because "pivot" is actually a different function in python (the R naming is so much better in this case...pivot_wider and pivot_longer. Python has "melt" and "pivot").

```

> We can achieve this effect in Python using the `.melt` function from the `pandas` package.

Reviewer 2: I think we decided to just use `method` instead of `.method`, so we should remove the prefixed `.` here and elsewhere

Vancouver, Calgary, and Edmonton columns into a single column called `region`,
and create a column called `mother_tongue` that contains the count of how many
Canadians report each language as their mother tongue for each metropolitan

> area:

Reviewer: avoid colon before code block

```{code-cell} ipython3
lang_mother_tidy = lang_wide.melt(

>     id_vars=["category", "language"],

Reviewer: This code block is different from the image above

Need to add a bold black comment for melissa to update the image (if the current code block is what we want to go with)

Author: I think it might be best to stick with the simpler code, so I will add a note

Author: Note added in 481

> **Note**: In the code above, the call to the
> `.melt` function is split across several lines. This is allowed in
> certain cases; for example, when calling a function as above, the input
> arguments are between parentheses `()` and Python knows to keep reading on
> the next line. Each line ends with a comma `,` making it easier to read.
> Splitting long lines like this across multiple lines is encouraged
> as it helps significantly with code readability. Generally speaking, you should

> > limit each line of code to about 80 characters.

Reviewer 2: Aren't we already introducing method chaining in the intro section chapter? In that case we can remove this from here I think.

Reviewer: I think it's fine (and often good) to repeat things like this -- but maybe we should link back to the explanation in Ch1?

> ```

Reviewer: Just above this: `lang_home_tidy.columns = [....]` is not explained at all. What is going on there? Do we need to do that any time we pivot? 

(also if we're renaming columns, we taught students how to do that in Ch 1 -- we should follow that example for consistency)

Reviewer: `reset_index` is also not explained at all

Author: added the text below. See what you think!

```{code-cell} ipython3

> lang_home_tidy.info()

Reviewer: holy crap this is a useful function!

We should absolutely teach `.info()` earlier. How about in a new section right before "Tidy data" (i.e., at the end of the "Data frames, series, and lists" section)

Reviewer: (and of course explain all the printouts you get with `.info()`

Author: added lines 655 and following to explain this. It doesn't seem like it needs its own heading? but we can easily add a heading if you prefer -- no strong opinions on my end

Reviewer 2: I don't think this needs its own heading but agree that we should introduce it early. One possible spot is when we talk about data frame column types and the `type` function. We mention the importance of having the same type in each every column there, but we do'n't show them how to see which data types are in a column, which `info` would remedy.

lang_long
```

What makes the data set shown above untidy?
In this example, each observation is a language in a region.
However, each observation is split across multiple rows:
one where the count for `most_at_home` is recorded,
and the other where the count for `most_at_work` is recorded.
Suppose the goal with this data was to
visualize the relationship between the number of
Canadians reporting their primary language at home and work.
Doing that would be difficult with this data in its current form,
since these two variables are stored in the same column.

> {numref}`fig:img-pivot-wider-table` shows how this data

Reviewer 2: It would be nice if this figure showed the params on separate lines as with the code block below and the image for `melt`

Reviewer 2: Also, it is unnecessary to specify `values` here since all remaining columns will be used for the values parameter. This is the same reason we omitted `value_vars` above, so I think it would make sense to omit it here too for consistency and simplicity. I think we can add a sentence mentioning that these parameter exists for both sections

```

First we’ll use `.melt` to create two columns, `region` and `value`,
similar to what we did previously.
The new `region` columns will contain the region names,
and the new column `value` will be a temporary holding place for the
data that we need to further separate, i.e., the
number of Canadians reporting their primary language at home and work.

```{code-cell} ipython3
lang_messy_longer = lang_messy.melt(
    id_vars=["category", "language"],
    var_name="region",
    value_name="value",
)

lang_messy_longer
```

Next we'll use `.str.split` to split the `value` column into two columns.

> How it works is that it takes a single string and splits it into multiple values

Reviewer: awk wording

> based on the character you tell it to split on. For example:

Reviewer: avoid colon before code block

lang_home_tidy
```

In the first step, note that we added a call to `reset_index`. When `pivot` is called with
multiple column names passed to the `index`, those entries become the "name" of each row that
would be used when you filter rows with `[]` or `loc` rather than just simple numbers. This
can be confusing... What `reset_index` does is sets us back with the usual expected behaviour
where each row is "named" with an integer. This is a subtle point, but the main take-away is that
when you call `pivot`, it is a good idea to call `reset_index` afterwards.

The second operation we applied is to rename the columns. When we perform the `pivot`
operation, it keeps the original column name `"count"` and adds the `"type"` as a second column name.

> Having two names for a column can be confusing! So we rename giving each column only one name.

Reviewer 2: I think it won't be clear to students why we do this when we previously used `rename`. I don't think there is a better way to flatten an index, [most other approaches ](https://datascientyst.com/flatten-multiindex-in-pandas/)would take more explaining. But it might make sense to explicitly say what we are doing here since it is different from what we taught them previously

> ```

Reviewer: again, lots of unexplained stuff going on above. `rename` and `drop`, `concat`, etc.

If we can simplify this code, we should

If we cannot (or can only to a small extent), we need to explain everything that is going on here in far more detail

Reviewer: It's often a good idea to "look at the output" of a partial operation, go "oh man there's something wrong now!" and then fix it. This is way less mystical for students, because they see individually why each step is happening.

Author: I broke this up into a few steps starting at 750. See if you think that is an improvement

Reviewer 2: An alternative to concat + rename is to assign directly to two columns in the dataframe:

```py
lang_messy_longer[["most_at_home", "most_at_work"]] = lang_messy_longer["value"].str.split("/", expand=True)
```

I think this is a bit easier to read myself (after showing them the output from the right hand expression once first), but if we want to make a point of showing concat because we think it is generally useful then I think the current way looks good.

Author: :+1: I like your suggestion @joelostblom: please feel free to change

```{code-cell} ipython3

> tidy_lang.info()

Reviewer: Do we need to print info here? or is the table print above enough?

:name: fig:img-separate
:figclass: caption-hack

Syntax for the `str.split` function.
```

We will do this in multiple steps. First, we create a new object
that contains two columns. We will set the `expand` argument to `True`
to tell `pandas` that we want to expand the output into two columns.

```{code-cell} ipython3
split_counts = lang_messy_longer["value"].str.split("/", expand=True)
split_counts
```
Since we only operated on the `value` column, the `split_counts` data frame
doesn't have the rest of the columns (`language`, `region`, etc.)
that were in our original dataframe. We don't want to lose this information, so
we will contatenate (combine) the original data frame with `split_counts` using
the `concat` function from `pandas`. The `concat` function *concatenates* data frames
along an axis. By default, it concatenates the data frames vertically along `axis=0` yielding a single
*taller* data frame. Since we want to concatenate our old columns to our
new `split_counts` data frame (to obtain a *wider* data frame), we will specify `axis=1`.

```{code-cell} ipython3
tidy_lang = pd.concat(
        (lang_messy_longer, split_counts),

>         axis=1,

Reviewer 2: ```suggestion
tidy_lang = pd.concat(
    [lang_messy_longer, split_counts],
    axis=1,
```

Same indentation as elsewhere and change to a list because we don't teach tuples I think?

Reviewer: agree!

We can see that this data now satisfies all three criteria, making it easier to

> analyze. But we aren't done yet! Notice in the table, all of the variables are

Reviewer 2: ```suggestion
analyze. But we aren't done yet! Although we can't see it in the data frame above, all of the variables are actually
```

The dtype is not shown in pandas as it is in dplyr

"object" data types. We can check this using the `info` method.
```{code-cell} ipython3
tidy_lang.info()
```

Object columns in `pandas` data frames are columns of strings or columns with
mixed types. In the previous example in the section on {ref}`pivot-wider`, the
`most_at_home` and `most_at_work` variables were `int64` (integer), which is a type of numeric data.
This change is due to the delimiter (`/`) when we read in this messy data set.
Python read these columns in as string types, and by default, `str.split` will
return columns with the `object` data type.

It makes sense for `region`, `category`, and `language` to be stored as an
`object` type. However, suppose we want to apply any functions that treat the
`most_at_home` and `most_at_work` columns as a number (e.g., finding rows
above a numeric threshold of a column).
That won't be possible if the variable is stored as a `object`.

> Fortunately, the `pandas.to_numeric` function provides a natural way to fix problems

Reviewer 2: There is also `infer_objects` and `convert_dtypes` which should do this automatically (but for some reason doesn't work here it seems... might still be worth a mention)

pd.Series(["Vancouver", "Toronto"]).isin(pd.Series(["Toronto", "Vancouver"]))
```

> ### Extracting rows above or below a threshold using `>` and `<`

Reviewer 2: Should we also mention `>=` and <=`?

Reviewer: yes

Reviewer: but no need for a special section (or title) for it. Just mention it in this section

glue("most_french", "{0:,.0f}".format(2669195))
```

We saw in the section on {ref}`filter-and` that
{glue:text}`most_french` people reported
speaking French in Montréal as their primary language at home.
If we are interested in finding the official languages in regions
with higher numbers of people who speak it as their primary language at home
compared to French in Montréal, then we can use `[]` to obtain rows
where the value of `most_at_home` is greater than
{glue:text}`most_french`.

```{code-cell} ipython3
official_langs[official_langs["most_at_home"] > 2669195]
```

This operation returns a data frame with only one row, indicating that when
considering the official languages,
only English in Toronto is reported by more people
as their primary language at home
than French in Montréal according to the 2016 Canadian census.

> ### Extracting rows using `query`

Reviewer 2: I wonder if it will be confusing that we are introducing a second way of doing the same thing... maybe we should add an explicit comment on that the other way is the default way? Or remove `query` unless we make heavy use of it later in the book

You can also extract rows above, below, equal or not-equal to a threshold using the
`query` method. For example the following gives us the same result as when we used
`official_langs[official_langs["most_at_home"] > 2669195]`.

```{code-cell} ipython3
official_langs.query("most_at_home > 2669195")
```

The query (criteria we are using to select values) is input as a string. This
comes in handy when we chain operations together.

(loc-iloc)=

> ## Using `loc[]` to filter rows and select columns.

Reviewer 2: To me it seems like everything that is done is this section could be done with just `[]` and there is no need to involve `loc` here. We can save `loc` for only those situation when we are both filtering rows and selecting columns, similar to what we did in the intro chapter.

which provides the ability to index with integers rather than the names of the columns.
For example, the column names of the `tidy_lang` data frame are
`['category', 'language', 'region', 'most_at_home', 'most_at_work']`.
Using `iloc[]`, you can ask for the `language` column by requesting the
column at index `1` (remember that Python starts counting at `0`, so the second item `'language'`
has index `1`!).

```{code-cell} ipython3
column = tidy_lang.iloc[:, 1]
column
```

You can also ask for multiple columns, just like we did with `[]`. We pass `:` before
the comma, indicating we want to retrieve all rows, and `1:` after the comma
indicating we want columns after and including index 1 (*i.e.* `language`).

```{code-cell} ipython3
column_range = tidy_lang.iloc[:, 1:]
column_range
```

The `iloc[]` method is less commonly used, and needs to be used with care.
For example, it is easy to
accidentally put in the wrong integer index! If you did not correctly remember
that the `language` column was index `1`, and used `2` instead, your code
would end up having a bug that might be quite hard to track down.

```{index} pandas.Series; str.startswith
```

+++ {"tags": []}

> ## Aggregating data with `agg`

Reviewer 2: ```suggestion
## Aggregating data
```

Since we cover more than just `agg` in this section

:name: tab:basic-summary-statistics
| Function | Description |
| -------- | ----------- |
| `count` | The number of observations (rows) |
| `mean` | The mean of the observations |
| `median` | The median value of the observations |
| `std` | The standard deviation of the observations |
| `max` | The largest value in a column |
| `min` | The smallest value in a column |
| `sum` | The sum of all observations |
| `agg` | Aggregate multiple statistics together |
| `describe` | a summary |
```

+++
+++

> **Note:** In `pandas` DataFrame, the value `NaN` is often used to denote missing data.
> By default, when `pandas` calculates summary statistics (e.g., `max`, `min`, `sum`, etc),
> it ignores these values. If you look at the documentation for these functions, you will
> see an input variable `skipna`, which by default is set to `skipna=True`. This means that
> `pandas` will skip `NaN` values when computing statistics.

### Calculating summary statistics on dataframes

What if you want to calculate summary statistics on an entire dataframe? Well,
it turns out that the functions in {numref}`tab:basic-summary-statistics`
can be applied to a whole data frame!

> For example, we can ask for the number of rows that each column has using `count`.

Reviewer 2: Hmmm, is this a useful thing to show? Doesn't data frames require that all column have the same number of rows so we would use `info` or `shape` to get the number of rows in the data frame instead?

```{code-cell} ipython3
region_lang.count()
```
Not surprisingly, they are all the same. We could also ask for the `mean`, but
some of the columns in `region_lang` contain string data with words like `"Vancouver"`
and `"Halifax"`---for these columns there is no way for `pandas` to compute the mean.

> So we provide the keyword `numeric_only=True` so that it only computes the mean of columns with numeric values. This

Reviewer 2: How annoying that pandas is starting to require this, I quite liked the convenient of not having to specify it... it seems like a reasonable default to only take the average of numerical columns... I wonder what their rationale is.

is also needed if you want the `sum` or `std`.
```{code-cell} ipython3
region_lang.mean(numeric_only=True)
```
If we ask for the `min` or the `max`, `pandas` will give you the smallest or largest number
for columns with numeric values. For columns with text, it will return the
least repeated value for `min` and the most repeated value for `max`. Again,
if you only want the minimum and maximum value for
numeric columns, you can provide `numeric_only=True`.
```{code-cell} ipython3
region_lang.max()
```
```{code-cell} ipython3
region_lang.min()
```

Similarly, if there are only some columns that you would like to get summary statistics of,

> we can first use `loc[]` and then ask for the summary statistic. An example of this is illustrated in {numref}`fig:summarize-across`.

Reviewer 2: ```suggestion
we can first use `[]` and then ask for the summary statistic. An example of this is illustrated in {numref}`fig:summarize-across`.
```

Later, we will talk about how you can use a more general function, `apply`, to also accomplish this.

```{figure} img/summarize/summarize.003.jpeg
:name: fig:summarize-across
:figclass: caption-hack

`loc[]` or `apply` is useful for efficiently calculating summary statistics on many columns at once. The darker, top row of each table represents the column headers.
```

Lets say that we want to know
the mean and standard deviation of all of the columns between `"mother_tongue"` and `"lang_known"`.

> We use `loc[]` to specify the columns and then `agg` to ask for both the `mean` and `std`.

Reviewer 2: ```suggestion
We use `[]` to specify the columns and then `agg` to ask for both the `mean` and `std`.
```

```{code-cell} ipython3

> region_lang.loc[:, "mother_tongue":"lang_known"].agg(["mean", "std"])

Reviewer 2: ```suggestion
region_lang["mother_tongue":"lang_known"].agg(["mean", "std"])
```

If we make the change I suggestion to this section above and only use loc for both filtering rows and selecting coluns at the same time

english_lang
```
We then wanted to add the populations of these cities as a column using `assign`
(Toronto: 5928040, Montréal: 4098927, Vancouver: 2463431,
Calgary: 1392609, and Edmonton: 1321426). We had to be careful to add those populations in the
right order, and it could be easy to make a mistake this way. An alternative approach, that we demonstrate here
is to (1) create a new, empty data frame, (2) use `assign` to assign the city names and populations in that
data frame, and (3) use `merge` to combine the two data frames, recognizing that the "regions" are the same.

We start by creating a new, empty data frame by calling `pd.DataFrame` and assigning the output
to `city_populations`. We then use `assign` to add the city names in a column called `"region"`
and their populations in a column called `"population"`.
```{code-cell} ipython3
city_populations = pd.DataFrame()  # empty data frame
city_populations = city_populations.assign(
  region=["Toronto", "Montréal", "Vancouver", "Calgary", "Edmonton"],
  population=[5928040, 4098927, 2463431, 1392609, 1321426]
)
city_populations
```
This new data frame has the same `region` column as the `english_lang` data frame. The order of

> the cities is different, but that is okay, we can use the `merge` function in `pandas` to say

Author: note that I added the introduction of `merge` here. It seems like a natural example but was not in the previous version, so please see what you think

Reviewer: more wrangling more better!

```

### Using `.assign` to modify columns

```{index} pandas.DataFrame; []
```

In the section on {ref}`str-split`,
when we first read in the `"region_lang_top5_cities_messy.csv"` data,
all of the variables were "object" data types.
During the tidying process,
we used the `pandas.to_numeric` function
to convert the `most_at_home` and `most_at_work` columns
to the desired integer (i.e., numeric class) data types and then used `[]` to overwrite columns.
But suppose we didn't use the `[]`,
and needed to modify the columns some other way.
Below we create such a situation
so that we can demonstrate how to use `.assign`
to change the column types of a data frame.
`.assign` is a useful function to modify or create new data frame columns.

```{code-cell} ipython3
lang_messy = pd.read_csv("data/region_lang_top5_cities_messy.csv")
lang_messy_longer = lang_messy.melt(
    id_vars=["category", "language"],
    var_name="region",
    value_name="value",
)
tidy_lang_obj = (
    pd.concat(
        (lang_messy_longer, lang_messy_longer["value"].str.split("/", expand=True)),
        axis=1,
    )
    .rename(columns={0: "most_at_home", 1: "most_at_work"})
    .drop(columns=["value"])

> )

Author: I am not sure this is critical to repeat. We could just pick up with `official_langs`? 

But I can also appreciate that it might be useful to revisit and see it all together
